src/tabs/Csg/canvas_holder.tsx

```diff
@@ -1,173 +1,165 @@
-          >
-            WebGL Context Lost
-          </h2>
-          <Spinner intent="warning" size={SpinnerSize.LARGE} />
-          <p
-            style={{
-              margin: `${STANDARD_MARGIN} 0px 0px 0px`,
-            }}
-          >
-            Your GPU is probably busy. Waiting for browser to re-establish connection...
-          </p>
-        </div>
-      </>
-    );
-  }
-}
+/* [Imports] */
+import { Spinner, SpinnerSize } from '@blueprintjs/core';
+import { IconNames } from '@blueprintjs/icons';
+import React from 'react';
+import { Core } from '../../bundles/csg/core.js';
+import StatefulRenderer from '../../bundles/csg/stateful_renderer.js';
+import type { RenderGroup } from '../../bundles/csg/utilities.js';
+import HoverControlHint from './hover_control_hint';
+import type { CanvasHolderProps, CanvasHolderState } from './types';
+import { BP_CARD_BORDER_RADIUS, BP_TAB_BUTTON_MARGIN, BP_TAB_PANEL_MARGIN, BP_TEXT_MARGIN, CANVAS_MAX_WIDTH } from '../common/css_constants.js';
+
+
+
+/* [Main] */
+export default class CanvasHolder extends React.Component<
+CanvasHolderProps,
+CanvasHolderState
+> {
+  private readonly canvasReference: React.RefObject<HTMLCanvasElement> = React.createRef();
+
+  private statefulRenderer: StatefulRenderer | null = null;
+
+  constructor(props: CanvasHolderProps) {
+    super(props);
+
+    this.state = {
+      isContextLost: false,
+    };
+  }
+
+  componentDidMount() {
+    console.debug(`>>> MOUNT #${this.props.componentNumber}`);
+
+    let { current: canvas } = this.canvasReference;
+    if (canvas === null) return;
+
+    let renderGroups: RenderGroup[] = Core
+      .getRenderGroupManager()
+      .getGroupsToRender();
+    let lastRenderGroup: RenderGroup = renderGroups.at(-1) as RenderGroup;
+
+    this.statefulRenderer = new StatefulRenderer(
+      canvas,
+      lastRenderGroup,
+      this.props.componentNumber,
+
+      () => this.setState({ isContextLost: true }),
+      () => this.setState({ isContextLost: false }),
+    );
+    this.statefulRenderer.start(true);
+  }
+
+  componentWillUnmount() {
+    console.debug(`>>> UNMOUNT #${this.props.componentNumber}`);
+
+    this.statefulRenderer?.stop(true);
+  }
+
+  // Only required method of a React Component. Returns a React Element created
+  // via JSX to instruct React to render a DOM node. Also attaches the
+  // canvasReference via the ref attribute, for imperatively modifying the
+  // canvas
+  render() {
+    return <>
+      <div
+        style={{
+          display: this.state.isContextLost ? 'none' : 'flex',
+          // Centre content when sidebar is wider than it
+          justifyContent: 'center',
+        }}
+      >
+        <div
+          style={{
+            display: 'flex',
+            flexDirection: 'column',
+            gap: BP_TAB_BUTTON_MARGIN,
+
+            marginRight: BP_TAB_PANEL_MARGIN,
+          }}
+        >
+          <HoverControlHint
+            tooltipText="Zoom in: Scroll up"
```

Comment:
For tooltips, would it be better to wrap the corresponding mouse actions in brackets?

Reply:
> Zoom in: (Scroll up)
> Zoom out: [Scroll down]
> Zoom to fit: {Double left-click}
> Pan: <Middle-click OR shift + left-click>

Unsure if I prefer the look of these options. I removed the `•` to go for something simpler and slightly clearer (hopefully).

It is worth noting that the hints are not buttons - the mouse actions are the current way to interact with the canvas.

Reply:
I see, I was under the assumption that they were buttons due to how they are placed and aligned. On that note, are there any plans to implement the buttons to actually perform the actions? Otherwise, it would seem more logical to just have a information box/tooltip with all the icons and hints in a single container.

Reply:
You are not the first person to think that. But adding click functionality to the hints does have a bit of a history. See also Cloud7050/modules#23 #214 #229.

Now that modules are able to import and use real BlueprintJS components, that does open up possibilities wrt consistent buttons. The slightly tricky part is that not all the current hints make sense as buttons.

Reply:
A single Blueprint hoverable Popover, as opposed to multiple icons and plaintext tooltips, could fit better for the purpose of explaining the controls, if no click functionality is planned. But since #229 would modify the current hints and change their usage, perhaps the UI can be further refined alongside/after that enhancement.